[feat] add documents for QA (#47)
dwarvesf/playbook: engineering/Testing/QA-best-practices.md

# QA best practises

## 1. Purpose of this document

This is a summary of QA practices Dwarvesv uses and recommends to be used. It is not supposed to be a detailed description and sometimes cannot fully be used of all tasks but as an overview of the most important QA processes and a list of good practices that should be used.

> ## 2. QA practices

Reviewer: ```suggestion
## QA practices
```

> Everybody in Dwarvesv team has QA responsibilities, even if there is a named QA manager or QA specialists. To Dwarvesv QA means three things.

Reviewer: ```suggestion
Everyone in Dwarves has QA responsibilities, even if there is a named QA manager or QA specialists. To us, QA means three things.
```

1.  The service runs as expected and is considered to be created with good practices.
2.  The service can be easily and cost efficiently maintained and operated.
3.  The service can be continuously improved and modified cost efficiently.

On high level QA work and practices can be divided into two processes.

1.  QA during agile sprints
2.  QA during deployment process

Furthermore there are other QA actions.

#### Test-Driven Development

Dwarvesv uses TDD and ATDD (Acceptance Test Driven Development) methods when applicable.

Method forces the implementation to follow architecture and consider modules that are used. Implementation starts by writing the automated test first and then implementing the functionality to pass the test.

#### Pair programming

Dwarvesv uses Pair programming method when applicable. This is very convenient way to share knowledge and experience about the project and software under development.

#### Code review

> Reviewing the code helps other team members to get information of certain functionality and gives a possibility to give feedback to responsible person and also ensures knowledge sharing between team members.

Reviewer: ```suggestion
Reviewing the code helps other team members to get information on certain functionality and gives a possibility to give feedback to the responsible person and also ensures knowledge sharing between team members.
```

#### Manual functional testing

> Manual testing is mostly done using Exploratory testing methodology and found errors are either fixed right away or prioritised and recorded to task/story/error management tool. Exploring the app or service can be started right after something functional is “ready”.

Author: In this case, I refer to using Exploratory alongside with all the other testing methods.

> # QA-best-practises

Reviewer: The guide from Futurice is good but we could carefully examine if we should take it to be our company standards. I will ask other QAs to help review the practices.

Author: I think it's just suggestion, we will align for each project instead of apply as the same to all.

> # QA-best-practises

Author: updated

## 1. Purpose of this document

This is a summary of QA practices Dwarvesv uses and recommends to be used. It is not supposed to be a detailed description and sometimes cannot fully be used of all tasks but as an overview of the most important QA processes and a list of good practices that should be used.

## 2. QA practices

Everybody in Dwarvesv team has QA responsibilities, even if there is a named QA manager or QA specialists. To Dwarvesv QA means three things.

1.  The service runs as expected and is considered to be created with good practices.
2.  The service can be easily and cost efficiently maintained and operated.

> 3.  The service can be continuously improved and modified cost efficiently

Reviewer 2: Missing dot at the end of the sentence, please add into it

Author: added

#### Code review

Reviewing the code helps other team members to get information of certain functionality and gives a possibility to give feedback to responsible person and also ensures knowledge sharing between team members.

#### Manual functional testing

Manual testing is mostly done using Exploratory testing methodology and found errors are either fixed right away or prioritised and recorded to task/story/error management tool. Exploring the app or service can be started right after something functional is “ready”.

Exploratory testing is a very powerful tool in e2e testing where the whole system is covered by testing. In the method tester goes beyond what can be defined in a test case, applies user-like thinking as well as tries to break the system by various error scenarios and is never “done” with testing.

#### Non functional testing - localization, usability, performance/load testing

Around the functional requirements and testing there are usually non-functional requirements which can be tested applying localisation, usability, performance and load testing. The needs and the tools are projects specific.

> For Performance/Load testing: LoadUI and J meter are actively in use to validate service capabilities in high traffic and to find possible service bottlenecks.

Reviewer 2: As far as I know, currently, from our company, we haven't done anything related to performance testing, I don't know if we should put it into here since we haven't worked with any of them before

Author: Removed
